### PR TITLE
dev/core#561 Convert grant search fields to use datepicker

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Converts  grant date search date files to use jcalendar

Before
----------------------------------------
screenshot 2018-12-03 11 28 34

After
----------------------------------------
screenshot 2018-12-03 11 36 29

Technical Details
----------------------------------------
https://github.com/civicrm/civicrm-core/pull/13210 is a pre-requisite

Takes some of the approach form https://github.com/civicrm/civicrm-core/pull/13013 (leaving out the default handling part & with some tweaks)

Comments
----------------------------------------
The grant fields are actually date fields not datetime - which would be more appropriate IMHO - so the change from date+time to date is appropriate here.

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@seamuslee001 ping

**Comment by seamuslee001** [comment]
Ok will look shortly @eileenmcnaughton

**Comment by seamuslee001** [comment]
@eileenmcnaughton can you rebase this now pls

**Comment by seamuslee001** [comment]
@eileenmcnaughton i note that in the change we have lost the date not set checkbox, is that a specific thing your changing here?

**Comment by eileenmcnaughton** [comment]
@seamuslee001 that checkbox seems to achieve the same thing as the little x you get with datepicker

**Comment by seamuslee001** [comment]
@eileenmcnaughton i think there is actually a functional difference because a clear date picker box means include dates and rows with no dates whereas looking at the implementation of a checkbox it looks like its about actually querying that the column is NULL https://github.com/civicrm/civicrm-core/pull/13211/files#diff-f5e872367f13e067d643ef634b73ba14R136

**Comment by eileenmcnaughton** [comment]
@seamuslee001 OK - I've returned the ugliness

**Comment by eileenmcnaughton** [comment]
@seamuslee001 argh - is that possible?

**Comment by seamuslee001** [comment]
@eileenmcnaughton not sure just assumed that it was the case with searches maybe not maybe only limited to contact based searches

**Comment by seamuslee001** [comment]
(*CiviCRM Review Template WORD-1.2*)



* General standards
    * ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) Pass
    * ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) Pass
    * ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) Pass
    * ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) Pass Tested the grant search form and confirmed that the change works
* Developer standards
    * ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) Pass
    * ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) Pass
    * ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) Pass
    * ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) Pass

**Comment by seamuslee001** [comment]
@eileenmcnaughton It looks like we have broken the tab on Advanced Search tho for grants

**Comment by eileenmcnaughton** [comment]
@seamuslee001 I looked a bit further & the date range stuff uses jcalendar :-( - worse it's used in lots of places. My currently thinking is that it's best to mature an approach for switching across (& for handling smart group data) before replacing instances elsewhere - as a bit of an 'unloved corner' grant feels like a good place to develop this approach - I've tinkered a little here but it is not complete from a formatting pov, a post process pov or a conversion pov.

**Comment by JoeMurray** [comment]
Tested on dmaster and functionally looks fine on Find Grants and on Advanced Search, including creating a Smart Group.

@monishdeb could you put in some love on this PR to advance the formatting, post process and conversion areas that @eileenmcnaughton mentions? I hope looking at the code the meaning of her points becomes clearer.

File: CRM/Contact/Form/Search.php [hunk 1]
```diff
@@ -902,4 +902,22 @@ public function getTitle() {
...
 
+  /**
+   * Do form value conversions based on form metadata.
+   */
+  public function convertFormValues() {
+    $metadata = $this->getSearchFieldMetadata();
+    foreach ($metadata as $fields) {
+      foreach ($fields as $field => $spec) {
+        if ($spec['type'] == CRM_Utils_Type::T_DATE) {
+          if (isset($this->_formValues[$field . '_low'])) {
```
Comment (seamuslee001) [inline CRM/Contact/Form/Search.php]: @eileenmcnaughton why only convert low? why not `_high` as well

File: templates/CRM/Core/DatePickerRange.tpl [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    </span>
+    <span class="crm-absolute-date-to">
+      {assign var=toName   value=$fieldName|cat:$to}
+      {$form.$toName.label}
+      {$form.$toName.html}
+    </span>
+  </span>
+  {literal}
+    <script type="text/javascript">
+      cj("#{/literal}{$relativeName}{literal}").change(function() {
```
Comment (seamuslee001) [inline templates/CRM/Core/DatePickerRange.tpl]: @eileenmcnaughton cj should be changed to i think $

File: CRM/Contact/Form/Search.php (hunk 1 above)
Comment (eileenmcnaughton) [reply CRM/Contact/Form/Search.php]: only because it's still incomplete

File: templates/CRM/Core/DatePickerRange.tpl (hunk 2 above)
Comment (colemanw) [reply templates/CRM/Core/DatePickerRange.tpl]: Ideally code like that gets enclosed in `CRM.$(function($) { ... });`

Comment (eileenmcnaughton) [reply templates/CRM/Core/DatePickerRange.tpl]: @colemanw fix pushed

File: CRM/Contact/Form/Search.php (hunk 1 above)
Comment (eileenmcnaughton) [reply CRM/Contact/Form/Search.php]: @seamuslee001 fix pushed

**Comment by eileenmcnaughton** [comment]
@seamuslee001 I've fixed this up & I'm happy it's working as far as it goes. I've decided to remove the patch that does the 'in-form' conversion in favour of an upgrade conversion & am working on that as a separate patch

**Comment by colemanw** [comment]
Thanks for the testing @seamuslee001. I left a note for @eileenmcnaughton on the other PR and will merge both once that's resolved.

**Pull request merged**